file-events/src/test/groovy/net/rubygrapefruit/platform/file/MovedDirectoriesFileEventFunctionsTest.groovy

```diff
@@ -0,0 +1,160 @@
+        def watchedDir = new File(parentDir, "removed")
+        watchedDir.mkdirs()
+        def removedFile = new File(watchedDir, "file.txt")
+        createNewFile(removedFile)
+        File removedDir = removedDirectory(watchedDir)
+        startWatcher(watchedDir)
+
+        when:
+        def directoryRemoved = removedDir.deleteDir()
+        // On Windows we don't always manage to remove the watched directory, but it's unreliable
+        if (!Platform.current().windows) {
+            assert directoryRemoved
+        }
+
+        def expectedEvents = []
+        if (Platform.current().macOs) {
+            expectedEvents << change(INVALIDATED, watchedDir)
+            if (ancestry == "watched directory") {
+                expectedEvents << change(REMOVED, watchedDir)
+            }
+        } else if (Platform.current().linux) {
+            expectedEvents << change(REMOVED, removedFile) << change(REMOVED, watchedDir)
+        } else if (Platform.current().windows) {
+            expectedEvents << change(MODIFIED, removedFile) << optionalChange(REMOVED, removedFile) << change(REMOVED, watchedDir)
+        }
+
+        then:
+        expectEvents expectedEvents
+
+        where:
+        ancestry                            | removedDirectory
+        "watched directory"                 | { it }
+        "parent of watched directory"       | { it.parentFile }
+        "grand-parent of watched directory" | { it.parentFile.parentFile }
+    }
+
+    @Issue("https://github.com/gradle/native-platform/issues/193")
+    def "can rename watched directory"() {
+        given:
+        def watchedDirectory = new File(rootDir, "watched")
+        watchedDirectory.mkdirs()
+        startWatcher(watchedDirectory)
+
+        when:
+        watchedDirectory.renameTo(new File(rootDir, "newWatched"))
+        waitForChangeEventLatency()
+        then:
+        if (Platform.current().linux) {
+            expectLogMessage(WARNING, Pattern.compile("Unknown event 0x800 for ${Pattern.quote(watchedDirectory.absolutePath)}"))
+        }
+        noExceptionThrown()
+    }
+
+    @Requires({ Platform.current().windows })
+    def "stops watching when path is moved"() {
+        given:
+        def watchedDir = new File(rootDir, "watched")
+        assert watchedDir.mkdirs()
+        def createdFile = new File(watchedDir, "created.txt")
+        startWatcher(watchedDir)
+
+        def renamedDir = new File(rootDir, "renamed")
+        assert watchedDir.renameTo(renamedDir)
+
+        when:
+        def droppedPaths = watcher.stopWatchingMovedPaths()
+        then:
+        droppedPaths == [watchedDir]
+
+        when:
+        assert createdFile.createNewFile()
+        then:
+        expectNoEvents()
+
+        when:
+        droppedPaths = watcher.stopWatchingMovedPaths()
+        then:
+        droppedPaths == []
+    }
+
+    @Requires({ Platform.current().linux || Platform.current().windows })
+    def "stops watching when parent of path is moved"() {
```

Comment:
Can you add a test here that on macOs, when we move a parent of the watched directory we receive file events for the watched directory? And when we then create a new directory at the location of the watched directory we receive events?